runtime/src/message_processor.rs

```diff
@@ -341,10 +348,19 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
         }
     }
     fn get_caller(&self) -> Result<&Pubkey, InstructionError> {
-        self.program_ids
+        self.invoke_stack
             .last()
+            .map(|frame| &frame.0)
             .ok_or(InstructionError::GenericError)
     }
+    fn remove_first_keyed_account(&mut self) {
+        let index = self.invoke_stack.len() - 1;
+        let last_frame = &mut self.invoke_stack[index];
+        last_frame.1 = &last_frame.1[1..];
```

Comment:
Tuple references are tough on the eyes, maybe make stack a vec of structs instead?

Reply:
Sure